clash-cores/src/Clash/Cores/Xilinx/DcFifo/Explicit.hs

```diff
@@ -0,0 +1,247 @@
+{-|
+Copyright  :  (C) 2022, Google Inc,
+License    :  BSD2 (see the file LICENSE)
+Maintainer :  QBayLogic B.V. <devops@qbaylogic.com>
+
+Support for the [FIFO Generator v13.2](https://docs.xilinx.com/v/u/en-US/pg057-fifo-generator).
+
+
+-}
+
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE QuasiQuotes #-}
+
+{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
+
+module Clash.Cores.Xilinx.DcFifo.Explicit
+  ( -- * Instantiating IP
+    dcFifo
+  , XilinxFifo(..)
+
+    -- * Customizing IP
+  , DcConfig(..)
+  , DcImplementation(..)
+  , ReadMode(..)
+  , defConfig
+
+    -- * Helper type aliases
+  , ResetBusy
+  , Full
+  , Empty
+  , DataCount
+  ) where
+
+import           Clash.Annotations.Primitive (Primitive (InlineYamlPrimitive))
+import           Clash.Explicit.Prelude
+import           Clash.Signal.Internal       (Signal (..))
+
+
+import qualified Data.Sequence               as Seq
+import           Data.String.Interpolate     (i)
+
+-- We want constructor names to correspond 1:1 to Xilinx options it ease
+-- blackbox implementation.
+{-# HLINT ignore "Use camelCase" #-}
+
+type ResetBusy = Bool
+type Full = Bool
+type Empty = Bool
+type DataCount n = Unsigned n
+
+data ReadMode
+  = Standard_FIFO
+  | First_Word_Fall_Through
+  deriving (Show, Generic)
+
+data DcImplementation
+  = Independent_Clocks_Block_RAM
+  | Independent_Clocks_Distributed_RAM
+  | Independent_Clocks_Builtin_FIFO
+  deriving (Show, Generic)
```

Comment:
I added these, but I was being over-eager. We only really test `Independent_Clocks_Block_RAM` - and the datasheet mentions some pretty crucial differences between these implementations. In the interest of time, I think we should just entirely remove this data type and hardcode it on `Independent_Clocks_Block_RAM` in the blackbox.